### PR TITLE
Python ESSA: Fix performance of PythonSsaSourceVariable.hasRefinementEdge()

### PR DESCRIPTION
**Pull request by markshannon**

### CODE REVIEW

File: python/ql/src/semmle/python/dataflow/SsaDefinitions.qll [hunk 1]
```diff
@@ -79,24 +79,12 @@ abstract class PythonSsaSourceVariable extends SsaSourceVariable {
     }
 
     override predicate hasRefinementEdge(ControlFlowNode use, BasicBlock pred, BasicBlock succ) {
+        test_contains(pred.getLastNode(), use) and
```
Comment (taus-semmle) [inline python/ql/src/semmle/python/dataflow/SsaDefinitions.qll]: Compared to what the code did previously, `test_contains` contains an additional `instanceof Expr` check. It's not immediately clear to me whether this affects the behaviour. Can you weigh in on this?

File: python/ql/src/semmle/python/dataflow/SsaDefinitions.qll [hunk 2]
```diff
@@ -79,24 +79,12 @@ abstract class PythonSsaSourceVariable extends SsaSourceVariable {
...
-        and
+        (pred.getAFalseSuccessor() = succ or pred.getATrueSuccessor() = succ) and
         /* There is a store to this variable -- We don't want to refine builtins */
-        exists(this.(Variable).getAStore()) and
-        /* There is at least one use or definition of the variable that is reachable by the test */
-        exists(ControlFlowNode n |
-            n = this.getAUse() or
-            this.hasDefiningNode(n) |
-            pred.(ConditionBlock).strictlyReaches(n.getBasicBlock())
-        )
```
Comment (taus-semmle) [inline python/ql/src/semmle/python/dataflow/SsaDefinitions.qll]: Do we know which of these changes -- using `test_contains` and removing this conjunct -- was responsible for the improvement? Was it both?

File: python/ql/src/semmle/python/dataflow/SsaDefinitions.qll (hunk 1 above)
Comment (markshannon) [reply python/ql/src/semmle/python/dataflow/SsaDefinitions.qll]: The last node in a branching block will always be an `Expr`. So this doesn't change behaviour, just makes it faster.

File: python/ql/src/semmle/python/dataflow/SsaDefinitions.qll (hunk 2 above)
Comment (markshannon) [reply python/ql/src/semmle/python/dataflow/SsaDefinitions.qll]: Both, but mainly this one, I think.

**Pull request merged**